### PR TITLE
Replace system's noSideEffect `proc`s by `func`s

### PR DESCRIPTION
**Pull request by AmjadHD**

### CODE REVIEW

**Comment by AmjadHD** [comment]
Why does changing:
`proc cmp*(x: string, y: string) {.noSideEffect.}` to
`func cmp*(x: string, y: string)`
triggers an error:
/home/runner/work/Nim/Nim/lib/system.nim(969, 6) Error: implementation of 'system.cmp(x: string, y: string) [declared in /home/runner/work/Nim/Nim/lib/system.nim(969, 6)]' expected

**Comment by metagn** [comment]
Because it's a forward declaration. Its implementation would also need `func`, as `func` is not *exactly* equal to `proc {.noSideEffect.}`.

File: lib/system/memalloc.nim [hunk 1]
```diff
@@ -1,5 +1,5 @@
 when notJSnotNims:
-  proc zeroMem*(p: pointer, size: Natural) {.inline, noSideEffect,
+  func zeroMem*(p: pointer, size: Natural) {.inline,
```
Comment (metagn) [inline lib/system/memalloc.nim]: This seems unintuitive to me, calling memory manipulation procs functions. Same for things that allocate above like `setLen` or `&`, or magic procs like `compiles`. Just wondering if it makes sense to use `func` for every noSideEffect proc. Maybe at least the math & comparison ones.

Comment (AmjadHD) [reply lib/system/memalloc.nim]: It seems unintuitive that a `func` is not _exactly_ equal to `proc {.noSideEffect.}`, despite the manual saying [so](https://nim-lang.github.io/Nim/manual.html#procedures-func).

Comment (AmjadHD) [reply lib/system/memalloc.nim]: Much of the procedures converted are magical even math and comparison procs.

Comment (juancarlospaco) [reply lib/system/memalloc.nim]: As a first step, fix the ones that are "direct" to fix IMHO. 🙂

**Comment by AmjadHD** [comment]
I give up.

**Pull request closed without merging**